adbc.h

```diff
@@ -1183,7 +1204,11 @@ AdbcStatusCode AdbcStatementExecutePartitions(struct AdbcStatement* statement,
 ///   driver.
 ///
 /// Although drivers may choose any name for this function, the
-/// recommended name is "AdbcDriverInit".
+/// recommended name is "AdbcDriverInit". If you use the recommended
+/// name, then driver managers will be able to automatically find
+/// the entrypoint function. Drivers may also expose the same function
+/// under a unique symbol so that multiple drivers can be statically
+/// linked into a single binary without any conflicts.
```

Comment:
Though, the definitions of Adbc* will still conflict; the only way to use them would be via the driver manager (and whatever linker trickery is necessary to prefer the driver manager's implementation of all such symbols)

Reply:
So maybe instead we should suggest either or? I'm thinking projects would allow a compile-time option that determines whether it uses the driver-manager compatible name or the unique name.

Reply:
At that point I'm not sure it needs to be documented here?

Reply:
In that case the options are basically:

- One driver per binary (driver manager or no): use the Adbc* names. If loaded via the driver manager, it will automatically load the driver via the default name, AdbcDriverInit. (Also, when creating the table of pointers in AdbcDriverInit, use private/namespaced names for each function instead of the Adbc* name because otherwise the dynamic linker may resolve those names to the driver manager's definition instead of yours.)
- Multiple drivers in one shared object, loaded via the driver manager: expose multiple distinct DriverInit names, which the user will have to manually configure. (Same caveat as before regarding the table of binaries, obviously.)
- Multiple drivers, all statically linked at the end: the current implementations do not support this, but no driver can define any Adbc* symbol (or you have to do whatever linker magic is necessary to avoid using those symbols). You must link to and use the driver manager.

Reply:
This dance mostly comes from also allowing you to directly link to and use a driver. If we dropped that and specified that you must always/only use the driver manager, we could avoid this (there would still be AdbcDriverInit, but we could say something like the user must always provide the name of the symbol, or the name of the symbol is derived from the name of the library, or something like that).

Reply:
If you have concerns or complaints, I'd like to address them or hear them 😅 

I was debating whether this is the right place initially...it could go in the formal docs, too, but I think here is alright. At least it keeps this relatively self-contained.

Reply:
> This dance mostly comes from also allowing you to directly link to and use a driver.

Wait there might be something I am missing: Should a driver define the symbols `AdbcStatementExecute()` and friends? Or should it only expose those functions as pointers within the result of `AdbcDriverInit()`?

Reply:
The drivers in this repo define both.

Reply:
IIRC, drivers like the test ones Dewey implemented only define the latter because they're only intended to be used that way. That's also fine.

Reply:
I'd also say that these 'constraints' are really just a natural consequence of a few requirements. If you want to dynamically load drivers, you have to have some entrypoint; if you want to directly link to drivers, you have to define certain names; if you want both, then you have to expose both, and now if you want to statically link multiple drivers, something has to give.

Reply:
Thanks, I think I understand better. Is there a design doc on how we arrived on that? I'm particularly curious about why direct linking is important.

Reply:
I don't think there was a formal design doc, just a decision that there was no need for a separate driver vs user API (just a different way of formulating the same viewpoint). Very early drafts did have a distinction, but I quickly realized I would just be copy pasting each function prototype.

Reply:
This _is_ different from JDBC/ODBC where (to my knowledge) you never use those APIs without the driver manager, in which case some of these questions would be moot. But the current state does make it easier to test and use the drivers (no need to all depend on the driver manager for development and testing, no need to implement the driver manager in order to do any development at all to start with)

Reply:
Makes sense. This is part of why I switched to implementing the driver manager *before* the implementation modules instead of after for Rust.

Though also in Rust the driver manager does the marshaling from Rust types to C types, which saves a lot of headaches. I'd bet it's the same for almost any other client language, except C. So makes me think direct linking is mostly useful for development purposes in C, and not something we expect users of the drivers to care about. What do you think of that?

> we could avoid this (there would still be AdbcDriverInit, but we could say something like the user must always provide the name of the symbol, or the name of the symbol is derived from the name of the library, or something like that).

Actually I think it would be very nice if we came up with a standard for deriving the Init symbol name from the library name. It could be used as a fallback if `AdbcDriverInit` isn't found. I'll create an issue for that.

Reply:
Neither Java nor Go require the manager.

Reply:
For Rust, I would've expected a 'native' `adbc::Driver` trait, I suppose. And then the driver manager's job (like Go) would be to wrap an unsafe FFI pointer into an implementation of that trait.

Reply:
That's basically what I've done. There's a concrete struct `FFI_AdbcDriver` that represents the struct as defined in the ADBC header. Then there's `AdbcDriver`, which is the safe wrapper around `FFI_AdbcDriver`.

I might be conflating some things when I say "driver manager". I'm basically calling the `AdbcDriver` struct a driver manager because it has a function `AdbcDriver::load()` which initializes a new instance from a shared library or init function.

But the API I've exposed so far in Rust always calls the function pointers on some `FFI_AdbcDriver`, and never tries to access any of the top-level `Adbc*` functions. I could add support for calling them maybe, but seems easier just to expose the `AdbcDriver::load()` path and tell people to always use that. Or if they really want to just call the `AdbcDriverInit` function themselves to create a `FFI_AdbcDriver` struct.

Reply:
Well yes that makes sense for FFI, but what about native Rust drivers?

Reply:
In what I've implemented so far, the only non-mangled symbol Rust drivers exposed is `AdbcDriverInit` (or whatever they implementor chooses to call it). So "native Rust drivers" aren't treated any different from those implemented in other languages. Does that answer your question? I'm not sure I fully understood it.

If a Rust application decided to statically link an AdbcDriver that was packaged as a Rust crate (either because it is implemented in Rust or else we wrapped another implementation into a crate), they would do something like:

```rust
use sqlite_adbc_driver::AdbcDriverInit;
use arrow_adbc::driver_manager::AdbcDriver;

let driver = AdbcDriver::load_from_init(AdbcDriverInit, ADBC_VERSION_1_0_0);
```

Reply:
I'm talking about Rust calling Rust drivers - there's no need to go through the driver manager or FFI in that case right?

Reply:
Basically, look at the Go implementation (or Java). There is no consideration of FFI or external drivers in the core library. It is an independent API. 

The FFI version is basically a special Go driver implementation that happens to use FFI in its implementation.

It sounds like the Rust API being designed is mostly/purely an FFI layer around the C API, and not a full ADBC standard for Rust.

Reply:
Ah then I must be misunderstanding some fundamental things here. I'll take a closer look at those.

Up till now, I've understood ADBC as a C API that could be implemented by C/C++/Go/Rust and consumed by any language. I've haven't considered what it means to be a standard *within* a language. But I suppose this interacts with some of my open questions for how packaging should work.

Reply:
It's a set of APIs in each language (unless that language only has bindings, like Python) - now, the C/Go/Java APIs are very similar to facilitate them being able to use each other's drivers. But the three 'core' APIs don't assume each other's existence as a prerequisite.

Reply:
Interesting. Here's a question: to what degree do we expect that we will share drivers vs each have their own implementation? i.e. will there be a FlightSQL ADBC driver for C, Go, and Java separately? And similarly with Postgres?

Reply:
FWIW I think the Rust implementation I've written isn't that far off here. With a little work we can probably transform this file into the "Rust ADBC API" https://github.com/wjones127/arrow-adbc/blob/rust-driver-manager/rust/src/interface.rs

Reply:
I've debated the same thing with myself.

I expect it'll come down to the specific realities of the implementation. Java is necessarily and unfortunately somewhat of an island (JNI bindings are somewhere in that list of things to do in the back of my head). C/Go/Rust (should) interop quite well and IMO it would be a win if we could build most/all drivers in Go/Rust and provide bindings to them for C/C++, provided we can solve packaging/distribution issues. There are specific benefits I'm seeking:

- Go/Rust make it easier to build heavily threaded code and/or deal with networking; even if Go is technically at a disadvantage vs C/C++, making it easier to get 90% of the way is better than struggling to get 70% of the way.
- Go/Rust are aimed more at building self contained binaries, which is maybe a tradeoff, but after dealing with C++ Protobuf packaging issues I think it would be worth examining that tradeoff.

That said, C/C++ are going to have easy access to libraries (libpq) and have a good amount of infrastructure set up (the validation suite, Googlebench + eventual Conbench, etc.). (But if we were to ditch libpq in favor of handrolling the client, then I would prefer Rust or Go, for instance, and we can always run the validation suite or C++ benchmark harness against Go or Rust.)

Reply:
Ok, that's what I figured (and what I was trying to drive at earlier)